### PR TITLE
feat(sort): custom series sort in study panel

### PR DESCRIPTION
**Pull request by IbrahimCSAE**
### Context

Allows defining custom displaySet sort in the customizationModule to be used for sorting the displaySets.

We are currently redesigning the study panel and the study browser. During this process, you can enable our undesigned component via the `experimentalStudyBrowserSort` flag. This will look like:

CleanShot 2024-06-20 at 16 09 57@2x

You can  also add your own sorting functions by utilizing the `customizationService` and adding the `studyBrowser.sortFunctions` key, as shown below:

```
customizationService.addModeCustomizations([
  {
    id: 'studyBrowser.sortFunctions',
    values: [{
      label: 'Series Images',
      sortFunction: (a, b) => {
        return a?.numImageFrames - b?.numImageFrames;
      },
    }],
  },
]);
```

### Related Issues and Discussions

- https://github.com/OHIF/Viewers/issues/4004
- https://github.com/OHIF/Viewers/issues/4076
- https://github.com/OHIF/Viewers/issues/3959
- https://github.com/OHIF/Viewers/issues/3933
- https://github.com/OHIF/Viewers/issues/4083
- https://github.com/OHIF/Viewers/pull/2285
- Community Discussion - [Sort instances/study by acquisition datetime](https://community.ohif.org/t/sort-instances-study-by-acquisition-datetime/1299/8)

### CODE REVIEW

**Review by sedghi — COMMENTED** [review]
The logic seems sound. 

However, we should avoid surprising users by introducing an unfinished UI feature. Therefore, please put it behind a flag that is disabled by default, perhaps named `experimentalStudyBrowserSort` or something similar. (we can just put it in default.js and netlify.js maybe) and add in the PR description and talk about it

Additionally, we need to include documentation in the FAQ explaining how users can add their custom sort functions. I'm tried of  repeatedly addressing such question, and i want to link to it

Overall great job

**Comment by sedghi** [comment]
I'm having second thoughts actually.

This should work at display set service level and not at the UI level. So upon sorting it should re-order displaysets and fire modified.

Panel study browser should not have a single line of change

**Comment by IbrahimCSAE** [comment]
> I'm having second thoughts actually.
> 
> This should work at display set service level and not at the UI level. So upon sorting it should re-order displaysets and fire modified.
> 
> Panel study browser should not have a single line of change

sounds good, let me sort that out monday

**Comment by lscoder** [comment]
Should this file should be updated?

https://github.com/OHIF/Viewers/blob/master/extensions/default/src/commandsModule.ts#L508-L511

**Comment by IbrahimCSAE** [comment]
> Should this file should be updated?
> 
> https://github.com/OHIF/Viewers/blob/master/extensions/default/src/commandsModule.ts#L508-L511

good catch

File: extensions/default/src/getCustomizationModule.tsx [hunk 1]
```diff
@@ -162,6 +165,26 @@ export default function getCustomizationModule({ servicesManager, extensionManag
...
+            {
+              label: 'Series Number',
+              sortFunction: (a, b) => {
+                return a?.SeriesNumber - b?.SeriesNumber;
+              },
+            },
+            {
+              label: 'Series Date',
+              sortFunction: (a, b) => {
+                const dateA = new Date(formatDate(a?.SeriesDate));
```
Comment (wayfarer3130) [inline extensions/default/src/getCustomizationModule.tsx:181]: If you are going to sort on series date, it HAS to include series time as well - there are cases where doing it wrong can result in patient harm.

File: modes/longitudinal/src/index.ts [hunk 2]
```diff
@@ -90,6 +90,18 @@ function modeFactory({ modeConfiguration }) {
       //   },
       // ]);
 
+      customizationService.addModeCustomizations([
```
Comment (wayfarer3130) [inline modes/longitudinal/src/index.ts:93]: Example customizations should be included in the basic test mode, not in longitudinal.

Comment (sedghi) [reply modes/longitudinal/src/index.ts:93]: I was just trying to see if it appends, but it did not

**Review by sedghi — APPROVED** [review]
Great

**Comment by niirdan** [comment]
good afternoon, the feature has been merged to master already?
 Is it available by adding a parameter in conf file?

**Comment by sedghi** [comment]
@niirdan Yes, take a look https://docs.ohif.org/faq/technical#how-can-i-change-the-sorting-of-the-thumbnail--study-panel--study-browser

**Comment by niirdan** [comment]
Thanks for the update!

‫בתאריך יום ד׳, 3 ביולי 2024 ב-15:56 מאת ‪Alireza‬‏ <‪
***@***.***‬‏>:‬

**Comment by niirdan** [comment]
I notice that sorting takes into consideration only cines , for example if
a study contains both images and cines it will sort only the cines and the
images will remain stacked together .
As a suggestion , In order to make this feature useful for cardiologists
they are looking for the sequence(acquisition date&time) of the study
material during their diagnosis .
To implement that , the stacking behaviour in this case is not the right
Idea.


‫בתאריך יום ד׳, 3 ביולי 2024 ב-16:00 מאת ‪Dan Nir‬‏ ***@***.***‬‏>:‬

**Comment by sedghi** [comment]
@niirdan Could you please elaborate on this? What do you mean by "stacking behavior"? Are you suggesting that the grouping is not done correctly? Which study are you using for your example? It would be helpful if you could use one of our studies from the demo server to explain the current behavior and the desired behavior.

**Comment by niirdan** [comment]
@sedghi for sure!
Cardiologists diagnosis for example stress-echo exams by comparing different breathing cases (In stress/rest etc..)
they acquire during the exam both single instance Img's and cine loops .
during the diagnosis they want to see the Chronological order of both img's and cines they took from the US because the sequence has medical reasons which are Important for them .
This is why the stacking of single img's (In This Case) is not an Idea presentation way. 
I managed to disable the stacking behavior - looks like that
list-with-sepimg
In this image you can see both a cine and image's , Which are not in the chronological ordered . 
The Dicom attribute which define the chronological order is acquisition date time , (0008,002A)

**Pull request merged**